miner/worker.go

```diff
@@ -676,31 +683,42 @@ func (w *worker) commitNewWork() {
 	txs := types.NewTransactionsByPriceAndNonce(w.current.signer, pending)
 	env.commitTransactions(w.mux, txs, w.chain, w.coinbase)
 
-	// Create the full block to seal with the consensus engine
-	fullState = env.state.Copy()
-	if fullBlock, err = w.engine.Finalize(w.chain, header, fullState, env.txs, uncles, env.receipts); err != nil {
-		log.Error("Failed to finalize block for sealing", "err", err)
-		return
-	}
+	w.finalize(uncles, w.fullTaskInterval, true)
+}
+
+// finalize runs any post-transaction state modifications, assembles the final block
+// and commits new work if consensus engine is running.
+func (w *worker) finalize(uncles []*types.Header, interval func(), update bool) error {
+	current := w.current
```

Comment:
What's the purpose of this variable closure?

Also minor nitpick, it's usually cleaner code style if you leave an empty line before a comment like the one below (`// Deep copy [...]`). It's hard on the eyes when code and comments mix tightly. If the previous line is a `}` only by itself, it's fine to not leave an extra empty line because you already have a visual separation (i.e. almost empty line).